src/Colopl/Spanner/Connection.php

```diff
@@ -443,4 +467,85 @@ protected function runQueryCallback($query, $bindings, Closure $callback)
 
         return $result;
     }
+
+    /**
+     * Returns current mode
+     *
+     * @return string
+     */
+    protected function getSessionNotFoundMode(): string
+    {
+        return $this->config['sessionNotFoundErrorMode'] ?? self::MAINTAIN_SESSION_POOL;
+    }
+
+    /**
+     * Handle "session not found" errors
+     *
+     * @template T
+     * @param  Closure(): T $callback
+     * @return T
+     * @throws InvalidArgumentException|NotFoundException|AbortedException
+     */
+    protected function withSessionNotFoundHandling(Closure $callback): mixed
+    {
+        $handlerMode = $this->getSessionNotFoundMode();
+        if (!in_array($handlerMode, [
+                self::MAINTAIN_SESSION_POOL,
+                self::CLEAR_SESSION_POOL,
+                self::THROW_EXCEPTION,
+            ])
+        ) {
+            throw new InvalidArgumentException("Unsupported sessionNotFoundErrorMode [{$handlerMode}].");
+        }
+
+        if ($handlerMode === self::THROW_EXCEPTION || $this->sessionPool === null) {
+            // skip handlers
+            return $callback();
+        }
+
+        try {
+            return $callback();
+        } catch (NotFoundException $e) {
+            // ensure if this really error with session
+            if ($this->causedBySessionNotFound($e)) {
+                if ($this->inTransaction()) {
+                    // if we inside transaction then throw abort exception
+                    throw new AbortedException(self::SESSION_NOT_FOUND_CONDITION, $e->getCode(), $e);
```

Comment:
I think we should throw this error after reconnecting.
That way we don't have to catch the AbortedException in the transaction code.

Reply:
Hi @taka-oyama 
this won't help - [Transaction](https://github.com/googleapis/google-cloud-php/blob/main/Spanner/src/Transaction.php) is tied to a session and stores it internally in private property. 
I tried to de-duplicate code in catch, please check 
https://github.com/colopl/laravel-spanner/pull/39/commits/df149c64df5c448f58c0e767b96465c2524c85a5

Reply:
So In that case, wouldn't it make sense to just throw the NotFoundException?

Wouldn't throwing AbortedException here [cause the transaction to retry](https://github.com/googleapis/google-cloud-php/blob/main/Spanner/src/Database.php#L864) with the same session?

Reply:
>So In that case, wouldn't it make sense to just throw the NotFoundException?
no, this will not work with reentrancy of withSessionNotFoundHandling

>Wouldn't throwing AbortedException here [cause the transaction to retry](https://github.com/googleapis/google-cloud-php/blob/main/Spanner/src/Database.php#L864) with the same session?

yes, I am going to reimplement with special exception class , something like TransactionRestartException

Reply:
Hi @taka-oyama 
I have implemented it without extra exception - just by handling ordinal QueryException

Reply:
Thank you for the fix!